core/src/engine/task/work_stealing_queue/task_queue.cpp

```diff
@@ -0,0 +1,82 @@
+#include "task_queue.hpp"
+#include <engine/task/task_context.hpp>
+#include "consumers_manager.hpp"
+
+namespace {
+thread_local userver::engine::Consumer* localConsumer = nullptr;
+}
+
+USERVER_NAMESPACE_BEGIN
+
+namespace engine {
+
+WorkStealingTaskQueue::WorkStealingTaskQueue(const TaskProcessorConfig& config)
+    : consumers_manager_(config.worker_threads),
+      consumers_count_(config.worker_threads),
+      consumers_order_(0) {
+  consumers_.reserve(consumers_count_);
+  for (size_t i = 0; i < consumers_count_; ++i) {
+    consumers_.emplace_back(this, &consumers_manager_, i);
+  }
+}
+
+void WorkStealingTaskQueue::Push(
+    boost::intrusive_ptr<impl::TaskContext>&& context) {
+  DoPush(context.get());
+  context.detach();
+}
+
+boost::intrusive_ptr<impl::TaskContext> WorkStealingTaskQueue::PopBlocking() {
+  boost::intrusive_ptr<impl::TaskContext> context{DoPopBlocking(),
+                                                  /* add_ref= */ false};
+  if (!context) {
```

Comment:
I see, `nullptr` as stop signal technique is not used here. Then this check should be `UASSERT` instead

Reply:
I use it because `Consumer` returns nullptr if stopped